src/Message/PurchaseRequest.php

```diff
@@ -1495,14 +1820,35 @@ public function getData()
                 'cc'         => $this->getWorkPhoneCountryPrefix(),
                 'subscriber' => $this->getWorkPhone(),
             ]);
+            // for optional threeDSRequestorAuthenticationInfo param
             $threeDSAuthInfo = array_filter([
                 'threeDSReqAuthData'      => $this->get3DsRequestAuthenticationMethodData(),
                 'threeDSReqAuthMethod'    => $this->get3DsRequestAuthenticationMethod(),
                 'threeDSReqAuthTimestamp' => $this->get3DsRequestAuthenticationTime(),
             ]);
+            // for optional DS_MERCHANT_EMV3DS param object
             $emv3DsParameters = array_filter([
+                // required parameters for v2
+                'browserAcceptHeader'                => $this->getBrowserAcceptHeader(),
+                'browserColorDepth'                  => $this->getBrowserColorDepth(),
+                'browserIP'                          => $this->getBrowserIP(),
+                'browserJavaEnabled'                 => $this->getBrowserJavaEnabled(),
+            //    'browserJavascriptEnabled'           => '', // conditional
+                'browserLanguage'                    => $this->getBrowserLanguage(),
+                'browserScreenHeight'                => $this->getBrowserScreenHeight(),
+                'browserScreenWidth'                 => $this->getBrowserScreenWidth(),
+                'browserTZ'                          => $this->getBrowserTZ(),
+                'browserUserAgent'                   => $this->getBrowserUserAgent(),
+                'notificationURL'                    => $this->getNotificationURL(),
+                'protocolVersion'                    => $this->getProtocolVersion(),
+                // Indicates whether the 3DSMethod has been executed. Values ​​accepted: Y= Successfully completed, N = Completed with errors, U = 3DSMethod not executed
+                'threeDSCompInd'                     => 'U',
+                // Type of request. Possible values: CardData, AuthenticationData, ChallengeResponse
+                'threeDSInfo'                        => 'CardData',
+                'threeDSServerTransID'               => $this->getTransactionId(),
+                // optional parameters for v2
                 'cardholderName'                     => $this->getCardholder(),
-                'email'                              => $card->getEmail(),
+                'email'                              => $card->getEmail(), // in docs as 'E-mail'
```

Comment:
Which docs, and this should be on the previous line

Reply:
Done